### PR TITLE
Add APIs for manual inter process context propagation

### LINKED ISSUE #334: Add manual inter-process context propagation capability to the public API
Maybe something like supported in the [OT inject-extract API](https://opentracing.io/guides/java/inject-extract/) - a key-value map that can be obtained from context for injection in client side and can be parsed into context in the server side

### PR DESCRIPTION
**Pull request by felixbarny**
closes elastic/apm-agent-java#334

### CODE REVIEW

**Comment by felixbarny** [comment]
This is an alternative design to #394

It has a much nicer API geared towards Java 8 but also provides fallbacks for Java 7.

For this to work on Java 7 we rely on the ability of a JVM to lazily link the method parameters so there is a small risk that on certain JVM implementations, the API does not work anymore for Java 7.

**Comment by felixbarny** [comment]
@kuisathaverat Pipeline Jenkins does not like that PR

**Comment by felixbarny** [comment]
TLDR;

adds these methods:

`ElasticApm`
* `Transaction startTransactionWithRemoteParent(Map<String, ? extends Iterable<String>> headers)`
* `Transaction startTransactionWithRemoteParent(Function<String, String> getFirstHeader)`
* `Transaction startTransactionWithRemoteParent(Function<String, String> getFirstHeader, Function<String, Iterable<String>> getAllHeaders)`

`Span`
* `void injectTraceHeaders(BiConsumer<String, String> addHeader)`
* `Map<String, String> getTraceHeaders()`

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java [hunk 1]
```diff
@@ -160,6 +178,25 @@ public static TraceContext with128BitId() {
         return AS_ROOT;
     }
 
+    @IgnoreJRERequirement
+    public static ChildContextCreator<Function<String, String>> fromHeaders() {
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java]: Risky in Java 7

**Comment by eyalkoren** [comment]
Jenkins please test this

**Comment by eyalkoren** [comment]
Jenkins test this please

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/FailSafeDeclaredMethodsCompiler.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * <p>
+ * One use case of this is to be able to instrument classes implementing {@code co.elastic.apm.api.Span},
+ * which refers to {@link java.util.function.BiConsumer}.
+ * </p>
+ * <p>
+ * Similar to {@link net.bytebuddy.dynamic.scaffold.MethodGraph.Compiler.ForDeclaredMethods} but eagerly resolves arguments and fails safe
+ * on exceptions in that it ignores methods causing exceptions.
+ * </p>
+ */
+public enum FailSafeDeclaredMethodsCompiler implements MethodGraph.Compiler {
```
Comment (felixbarny) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/FailSafeDeclaredMethodsCompiler.java]: not actually needed for this change but might still be relevant

File: apm-agent-api/src/main/java/co/elastic/apm/api/ElasticApm.java [hunk 3]
```diff
@@ -83,6 +84,60 @@ private static Object doStartTransaction() {
...
+     * </p>
+     * <p>
+     *
+     * @param headerExtractor a function which receives a header name and returns the fist header with that name
+     * @param headersExtractor  a function which receives a header name and returns all headers with that name
+     * @return the started transaction
+     * @since 1.3.0
+     */
+    @Nonnull
+    public static Transaction startTransactionWithRemoteParent(HeaderExtractor headerExtractor, HeadersExtractor headersExtractor) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-api/src/main/java/co/elastic/apm/api/ElasticApm.java]: even for protocols which do support multiple headers, we want to have the ability to get a single value for a header in case we know it can only have a single value, like `traceparent`.

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/ElasticApmApiInstrumentation.java [hunk 4]
```diff
@@ -68,6 +68,31 @@ private static void doStartTransaction(@Advice.Return(readOnly = false) Object t
...
+        private static void doStartTransaction(@Advice.Return(readOnly = false) Object transaction,
+                                               @Advice.Argument(0) MethodHandle getFirstHeader,
+                                               @Advice.Argument(1) @Nullable Object headerExtractor,
+                                               @Advice.Argument(0) MethodHandle getAllHeaders,
+                                               @Advice.Argument(3) @Nullable Object headersExtractor) throws Throwable {
+            if (tracer != null) {
+                if (headerExtractor != null) {
+                    final String traceparentHeader = (String) getFirstHeader.invoke(headerExtractor, TraceContext.TRACE_PARENT_HEADER);
+                    transaction = tracer.startTransaction(TraceContext.fromTraceparentHeader(), traceparentHeader);
+                } else {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/ElasticApmApiInstrumentation.java]: now implemented. If headerExtractor is null, headersExtractor is used unless it is also null
